elliptic-curve/src/group.rs

```diff
@@ -0,0 +1,114 @@
+//! Algebraic group traits vendored from the `group` crate
+
+// Copyrights in the "group" library are retained by their contributors. No
+// copyright assignment is required to contribute to the "group" library.
+//
+// The "group" library is licensed under either of
+//
+// * Apache License, Version 2.0, (see ./LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0)
+// * MIT license (see ./LICENSE-MIT or http://opensource.org/licenses/MIT)
+//
+// at your option.
+//
+// Unless you explicitly state otherwise, any contribution intentionally
+// submitted for inclusion in the work by you, as defined in the Apache-2.0
+// license, shall be dual licensed as above, without any additional terms or
+// conditions.
+
+// TODO(tarcieri): switch to the upstream crate. Blocked on `no_std` support:
+// <https://github.com/zkcrypto/group/pull/9>
+
+use core::fmt;
+use core::iter::Sum;
+use core::ops::{Add, AddAssign, Mul, MulAssign, Neg, Sub, SubAssign};
+use ff::PrimeField;
+use rand_core::{CryptoRng, RngCore};
+use subtle::Choice;
+
+/// A helper trait for types with a group operation.
+pub trait GroupOps<Rhs = Self, Output = Self>:
+    Add<Rhs, Output = Output> + Sub<Rhs, Output = Output> + AddAssign<Rhs> + SubAssign<Rhs>
+{
+}
+
+impl<T, Rhs, Output> GroupOps<Rhs, Output> for T where
+    T: Add<Rhs, Output = Output> + Sub<Rhs, Output = Output> + AddAssign<Rhs> + SubAssign<Rhs>
+{
+}
+
+/// A helper trait for references with a group operation.
+pub trait GroupOpsOwned<Rhs = Self, Output = Self>: for<'r> GroupOps<&'r Rhs, Output> {}
+impl<T, Rhs, Output> GroupOpsOwned<Rhs, Output> for T where T: for<'r> GroupOps<&'r Rhs, Output> {}
+
+/// A helper trait for types implementing group scalar multiplication.
+pub trait ScalarMul<Rhs, Output = Self>: Mul<Rhs, Output = Output> + MulAssign<Rhs> {}
+
+impl<T, Rhs, Output> ScalarMul<Rhs, Output> for T where T: Mul<Rhs, Output = Output> + MulAssign<Rhs>
+{}
+
+/// A helper trait for references implementing group scalar multiplication.
+pub trait ScalarMulOwned<Rhs, Output = Self>: for<'r> ScalarMul<&'r Rhs, Output> {}
+impl<T, Rhs, Output> ScalarMulOwned<Rhs, Output> for T where T: for<'r> ScalarMul<&'r Rhs, Output> {}
+
+/// This trait represents an element of a cryptographic group.
+pub trait Group:
+    Clone
+    + Copy
+    + fmt::Debug
+    + fmt::Display
+    + Eq
+    + Sized
+    + Send
+    + Sync
+    + 'static
+    + Sum
+    + for<'a> Sum<&'a Self>
+    + Neg<Output = Self>
+    + GroupOps
+    + GroupOpsOwned
+    + ScalarMul<<Self as Group>::Scalar>
+    + ScalarMulOwned<<Self as Group>::Scalar>
+{
+    /// Scalars modulo the order of this group's scalar field.
+    type Scalar: PrimeField;
```

Comment:
I feel like `Field` would be more accurate here; or, if you have `PrimeField`, the trait should be called `PrimeOrderGroup`. But since this is copied from `group` (and will eventually be replaced by its use), I should probably raise this question there.

Reply:
As it were, [`PrimeField`](https://docs.rs/ff/0.7.0/ff/trait.PrimeField.html) is from the [`ff`](https://github.com/zkcrypto/ff) crate (which also has a [`Field`](https://docs.rs/ff/0.7.0/ff/trait.Field.html) trait).

Reply:
I'm just saying, it seems logical to me to have `Group::Scalar: Field` and `PrimeGroup::Scalar: PrimeField`, but evidently the authors of `group` followed some different logic, perhaps better suited for ECC specifics.

Reply:
Aah, I see what you're saying.

I can't speak to why it is the way it is either, but if you're curious, here's `CofactorGroup`: https://github.com/zkcrypto/group/blob/master/src/cofactor.rs